rfd/00xx-pod-access-requests.md

```diff
@@ -0,0 +1,598 @@
+allow users to request access to specific Pods or Teleport administrators to
+limit the access that a particular user has to certain Pods.
+
+### Related issues
+
+- TBD
+
+## Why
+
+Teleport does not manage individual Kubernetes RBAC rules assigned to users or roles.
+Instead, Teleport allows users to impersonate a defined set of Kubernetes RBAC
+principals - groups or users - existing in the target cluster. Each user request
+will include the associated principals as impersonation headers, and Kubernetes
+RBAC will use them to control whether the request can be allowed.
+
+To change the user's permissions, you must create or edit a Kubernetes RBAC role
+in the Kubernetes cluster and associate it with an RBAC principal. Afterward, you
+append it into the Teleport user properties and the user can impersonate it.
+Currently, Teleport appends the RBAC principals to Teleport Roles or to the user
+itself. It means that if a user wants temporary access to a single Kubernetes POD,
+the cluster administrator has to create a Kubernetes RBAC Role, assign it to a
+Kubernetes RBAC principal and create a role with the principal. After that, the user
+has to request access to the temporary role. Once someone approves the role access
+request, the user can impersonate it.
+
+In addition to the complex process, limiting access to pods using Kubernetes RBAC
+has restrictions because Kubernetes RBAC does not support controlling access resources
+using pattern matching.
+The RBAC Role must include the resource name to grant access, but it requires
+knowing the Pod name in advance. In many cases, it is impossible to know them
+because Pods generated from Kubernetes objects like Deployments include random
+strings in their names and change each time the Pod is deleted or relocated from
+one node to the other.
+
+This RFD proposes an extension to Teleport RBAC where it will support
+requesting access to Pods - including Pods names following a pattern - without
+relying on requesting access to Teleport Roles.
+
+## Role definition
+
+This RFD proposes an alternative to limit or extend user access to certain pods
+in a Kubernetes cluster from Teleport.
+
+To achieve this, it proposes changes to Teleport's Roles to introduce a new
+Role version - `v6` - that extends the `v5` to include a field `kubernetes_pods` 
+where it's possible to specify the Pod names to allow access.
+
+```yaml
+kind: role
+version: v6
+metadata:
+  name: my-kube-role
+spec:
+  allow:
+    kubernetes_labels:
+      "*": "*"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_pods: ["A","B","C"]
+... other v5 fields
+```
+
+By default, `kubernetes_pods` is empty and will reject any access to Pods within
+the allowed clusters. To allow access to all Pods, the Role must include
+`kubernetes_pods: ["*"]`.
+
+Restricting access to Pods will only apply to Kubernetes clusters that satisfy
+the `kubernetes_labels` values. This means that the following role:
```

Comment:
Hey @nklaassen,

Good feedback is always welcome!

When a user dynamically requests access to a pod(s), it follows the exact same method you designed for normal resource access requests.
My implementation works as following:
- `tsh request create --resource /<cluster>/pod/<kube_cluster>/<pod_namespace>/<pod_name>`
This request is encoded using `ResourceId` in the following way:
```
types.ResourceId{
   Cluster: "<cluster>",
   Kind: "pod",
   Name: "<kube_cluster>", //does not support wildcards 
   SubResource: "<pod_namespace>/<pod_name>" // both support wildcards matching
}
```

We encode the ResourceId into the user's certificate and we evaluate it at runtime. When Kind is "pod", we need to allow access to the kube_cluster as well and we validate if the pod name is allowed. The resource name is not encoded in the role.

Regarding the `kubernetes_pod` requirement in the role spec, I do imagine the following cases.

1. Admins want to limit access to pods using wildcards or limit access to namespaces without expiration and without creating several Kubernetes RBAC roles.
2. Admins want to limit what users can search using `search_as_role` . I do imagine cases where only a subset of pods are searchable for a certain user/team.

It's not required for dynamic access requests but it's a nice addition tbh.

Reply:
Okay that is making a lot more sense now, thanks for clarifying it for me.

I think it could be made more clear in this doc that `kubernetes_pods` in the role spec is not actually required to implement the pod access requests, if I'm understanding that correctly. And maybe I'm not aware of some motivation for this feature, but my two cents would be to take it out if possible.

Introducing a breaking change to roles and a new major version is adding a lot of scope to this design when I think we could accomplish almost everything here without it. It is also introducing a new concept to our RBAC system, currently if no role has a `deny` for an action, any single role with a matching `allow` is enough, and the access check function actually returns early. With this we will have to check every role for any pod access, so it is a different conceptual model, we can no longer say to users "if any one role allows the action, it is allowed" there will be caveats.

We can always add that feature separately in the future if we want to, but I think you will have a much easier time with pod access requests if you can reduce the scope as much as possible. That's just my thoughts on it.

Reply:
@nklaassen 

Sorry for the delay but I missed the notification.

I do agree with you that we do not need the `kubernetes_pods` to implement the Pod Access requests, but from my discussions with @r0mant, this feature must include the Pod RBAC access control (static) + Pod Access requests.
The feature is described in several places with different names and maybe the one I choose for the RFD naming is not the best. I changed the RFD and the PR's name to present a more descriptive name for the feature we want to implement.

Regarding the breaking change: `RoleV6` will be fully compatible with `RoleV5` features and its protobuf will also be compatible so `pre-RoleV6` clients are compatible with the `RoleV6` Auth server.  It requires that the Auth Server downgrades the Role version for clients that do not support version V6 - similarly to the code introduced from `RoleV4` to `RoleV5` code, and in those cases, this feature won't work as expected since the filtering is done in the Kubernetes Service, i.e. costumers must upgrade their services to `>=12.0.0`.

For any role  `<v6`, Teleport will automatically populate `kubernetes_pods` with a wildcard in order to keep the current behavior and to avoid users updating every role in the cluster. `kubernetes_pods` will only be effective for roles with `version: v6`.

Regarding the changes required to Teleport RBAC, I do agree with you that this feature changes the RBAC logic. One possible solution for this is to compute the `kubernetes_groups` and `kubernetes_users` allowed at runtime. Currently, Auth server grabs them when a user requests a cert and embeds them into `tlsca.Identity` based on `kubernetes_labels` matching. 
What we can do is for the Pod endpoint re-compute them based on the request details and forward them to the Kubernetes cluster as Impersonation headers. For the Pod endpoints, Teleport will match the cluster RBAC and Pod RBAC and accumulate the Kubernetes Principals that fulfill the requirements before sending. This removes extra roles that do not match a certain pod and allow users to use different Kubernetes RBAC principals when reaching different pods. An issue will also be present when listing pods since we do not know which Impersonated headers grated access to each resource.